### PR TITLE
Kernel<-Fusion Step 4 - Kernel Inherits Fusion

### PR DESCRIPTION
**Pull request by csarofeen**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 1]
```diff
@@ -156,11 +156,6 @@ bool hasBlockSync(const Expr* expr, const ThreadPredicateMap& pred_map) {
   }
 
   auto tv = getTvOutput(expr);
-  if (tv->isKirStmt()) {
-    tv = tv->fuserTv();
-    expr = tv->definition();
-  }
-  TORCH_INTERNAL_ASSERT(expr != nullptr);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:163]: 🎉

File: torch/csrc/jit/codegen/cuda/ir_interface_nodes.h [hunk 2]
```diff
@@ -416,9 +408,6 @@ class TORCH_CUDA_CU_API TensorView : public Val {
   MemoryType memory_type_ = MemoryType::Local;
   SwizzleType swizzle_type_ = SwizzleType::NoSwizzle;
   std::vector<IterDomain*> axes_to_swizzle_;
-
-  // TODO: Remove, only used for lowering
-  const TensorView* fuser_tv_ = nullptr;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:421]: 🎉

File: torch/csrc/jit/codegen/cuda/executor.h [hunk 3]
```diff
@@ -193,6 +191,8 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
   static int fusion_id_counter_;
 
   GpuLower lowered_;
+  // Copy of lowered_->kernel()->as<Fusion>()
+  Fusion* fusion_;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/executor.h]: Please add ` = nullptr` just in case.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 4]
```diff
@@ -117,11 +115,12 @@ void FusionExecutor::debugCompileFusionFromStr(
               << std::endl;
   }
 
-  setUsedTVs();
+  lowered_ = GpuLower(fusion);
+  const auto kernel = lowered_.kernel();
+  fusion_ = lowered_.kernel()->as<Fusion>();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: nit: no need to have `->as<Fusion>()`.

File: torch/csrc/jit/codegen/cuda/fusion.h [hunk 5]
```diff
@@ -232,12 +226,15 @@ class TORCH_CUDA_CU_API Fusion final : public IrContainer {
...
 
   //! Register expr with this fusion.
   //! When we register an expression, we want to update the dependency tracking
-  //! of Vals. We add expr to our general expr_set_,
-  void registerExpr(Expr* expr) override;
+  //! of Vals. If this container is a not a Kernel, it will remove previous
+  //! definitions of outputs and register this Expr as the definition. Otherwise
+  //! will update definition if not previously set, but will not remove old
+  //! definitions.
+  virtual void registerExpr(Expr* expr);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/fusion.h]: Shouldn't this have `override`?

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 6]
```diff
@@ -412,18 +558,22 @@ void GpuLower::lower() {
       active_gpu_lower = nullptr;
     }
   } lower_guard(this);
+  // Copy fusion into a new kernel for processing
+  kernel_ = std::make_unique<kir::Kernel>(fusion);
+  // Alias the fusion kernel caries around as a view of itself.
+  fusion_ = kernel_->as<Fusion>();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp]: nit: No need to have `as<Fusion>()`.

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 7]
```diff
@@ -266,11 +266,110 @@ class KIRCleaner : public OptOutDispatch {
...
+}
 } // namespace
 
-void GpuLower::replaceSymbolicSizes() {
-  FUSER_PERF_SCOPE("GpuLower::Lower::replaceSymbolicSizes");
+class ValReplacementMutator : private OptOutMutator {
+ public:
+  ValReplacementMutator(
+      Fusion* fusion,
+      std::unordered_map<Val*, Val*> replacement_map)
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp]: nit: make it `const &`.

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 8]
```diff
@@ -397,9 +495,57 @@ void GpuLower::collectPaddedParallelDims() {
...
+    return inserter.exprs_;
+  }
+
+ private:
+  //! Insert Sync nodes at the end of a given for-loop when a WAR
+  //! hazard may happen.
+  UnaryOpInserter(const std::vector<Expr*>& exprs) {
+    kir::ExprMutator::traverseAndInsert(exprs);
+  }
+
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp]: I see compilation warnings here. Please insert `using kir::ExprMutator::handle;`

**Review by naoyam — APPROVED** [review]
Just some minor comments. Looks like there's a common pattern like `kernel->as<Fusion>()`. Now that `Kernel` is a subclass of `Fusion`, the pointer cast shouldn't be required.

File: torch/csrc/jit/codegen/cuda/executor.h (hunk 3 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.h]: Pushing this to https://github.com/csarofeen/pytorch/pull/1390

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp]: Pushing cleanup of all instances of this to https://github.com/csarofeen/pytorch/pull/1390

File: torch/csrc/jit/codegen/cuda/lower2device.cpp (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower2device.cpp]: Pushing all cleanup instances of this to https://github.com/csarofeen/pytorch/pull/1390

File: torch/csrc/jit/codegen/cuda/lower2device.cpp (hunk 7 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower2device.cpp]: Need to push this to next PR

File: torch/csrc/jit/codegen/cuda/lower2device.cpp (hunk 8 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower2device.cpp]: Need to push this to next PR

**Pull request merged**